libavcodec/vvc/vvc_ctu.h

```diff
@@ -96,6 +96,7 @@ typedef struct TransformBlock {
     int bd_offset;
 
     int *coeffs;
+    int16_t *pixels;
```

Comment:
why do we need pixels? can we just change coeffs to int16_t?

Reply:
If `sps_extended_precision_flag = 1`, then the transform coefficients are `bit_depth + 6` bits each.

Reply:
Note that if `bit_depth <= 10`, `sps_extended_precision_flag` must be 0. For now, I think it is easiest to treat transform coeffs as capable of fitting in a `int16_t` for `bit_depth <= 10` and not for `bit_depth > 10`, as we already have the mechanism to use different functions depending on bit-depth. This is what the current code is already doing. In the long term, yes I agree it will be better to change based on `sps_extended_precision_flag`, but as there are few > 10-bit videos which do not use the range extension I don't think it needs to be a priority.

Re. `coeffs`/`pixels`, this is the same problem I was trying to get at in https://github.com/ffvvc/FFmpeg/pull/117#issuecomment-1682161165. I'm not quite sure why you suggest `uint8_t *` — is this just meant as an arbitrary type which will be reinterpreted according to when an `int` or `int16_t` is needed? If that's the case I think `char` may be more fitting.

Aside: I nearly have the 32x32 transform working, there is just a single conformance test failing due to an overflow from a very large transform coefficient.

Reply:
> Re. coeffs/pixels, this is the same problem I was trying to get at in https://github.com/ffvvc/FFmpeg/pull/117#issuecomment-1682161165. I'm not quite sure why you suggest uint8_t * — is this just meant as an arbitrary type which will be reinterpreted according to when an int or int16_t is needed? If that's the case I think char may be more fitting.

sizeof(char) is not always 8, according to spec. it only can make sure sizeof(char) is >= 8. https://en.cppreference.com/w/cpp/language/types

we can define a new type coeff like this

```
#if BIT_DEPTH > 10 && EXTEND_PRECISION
 typedef Coeff int32_t;
#define TR_FUNC(f) FUNC(f##_ep)
#else 
 typedef Coeff int16_t;
#define TR_FUNC(f) FUNC(f)
#endif

static void TR_FUNC(transform_bdpcm)(uint8_t *_coeffs, const int width, const int height,
    const int vertical, const int log2_transform_range)
{
          Coeff *coeff = (Coeff*)_coeffs
...
}
```

Reply:
Here's something relevant I've found while getting 88f8a7dd21b91c20a9db28fe5c4788a9d05de4ce working. The residual samples produced by the transform process may be bigger than the eventual reconstructed pixels. For example, the 10-bit 32x32 DCT-II may produce residuals which are greater than `(1 << 10 - 1) - 1` and even greater than `INT16_MAX`. The value is only clipped _after_ being summed with the predicted sample. I think we may then need to use a larger type for the output of the transforms. I am not sure quite what the bound is on the residual samples, but I think `int32_t`s should be sufficient.

At the moment, this is manifesting in an intermittent failure for the ITX checkasm of `ff_vvc_inv_dct2_dct2_32x32_10_avx2`. In the case the output is so large it overflows the `int16_t` pixel type, the C implementation wraps around whereas the assembly implementation saturates, so the test fails. The patch below fails reliably.

<details>
<summary><b>Patch for tests/checkasm/vvc_itx.c to set all transform coefficients to maximum value</b></summary>

```diff
diff --git a/tests/checkasm/vvc_itx.c b/tests/checkasm/vvc_itx.c
index f403c4f1ac..17b18bcc03 100644
--- a/tests/checkasm/vvc_itx.c
+++ b/tests/checkasm/vvc_itx.c
@@ -35,7 +35,7 @@
         int k;                                                                  \
         for (k = 0; k < size; ++k) {                                            \
             uint32_t r = rnd();                                                 \
-            int32_t a = min + r / (UINT32_MAX / (max - min + 1) + 1);           \
+            int32_t a = min + UINT32_MAX / (UINT32_MAX / (max - min + 1) + 1);  \
             AV_WN32A(buf0 + k, a);                                              \
             AV_WN32A(buf1 + k, a);                                              \
         }                                                                       \
```

</details>

Reply:
> Rather than reinterpreting the uint8_t parameter, can we not just use Coeff in the signature? Would need some extra work in vvcdsp.h but probably beneficial in the long run.
Not possible. Coeff will defined to uint16_t and uint32_t in condition. we can't not use it as a signature.

thank you

Reply:
>The value is only clipped after being summed with the predicted sample
If it's only internal overflow(in the function), maybe you can use int32_t as an internal type. 

>At the moment, this is manifesting in an intermittent failure for the ITX checkasm of ff_vvc_inv_dct2_dct2_32x32_10_avx2
why overflow only happens at 32x32? will  16x16 .. 2x2 has the same issue? will 8 bits has the same issue?

Reply:
> > Rather than reinterpreting the uint8_t parameter, can we not just use Coeff in the signature? Would need some extra work in vvcdsp.h but probably beneficial in the long run.
> 
> Not possible. Coeff will defined to uint16_t and uint32_t in condition. we can't not use it as a signature.
> 
> thank you

Ah I see. dav1d uses their coeff/pixel types in their signatures, and I thought this was a template which was instantiated for each bit depth, but it looks as though it is only a build-time selection.

I am still not sure about the `uint8_t *` type though — if you were to do any manipulations on `coeffs` as `uint8_t`s it would cause corruption. Maybe a `void *` to make it more explicit that the array will need to be interpreted according to the bit depth and `sps_extended_precision_flag`?

Reply:
> > The value is only clipped after being summed with the predicted sample
>
> If it's only internal overflow(in the function), maybe you can use int32_t as an internal type.

Unfortunately this summing does not take place in the transforms but instead later in `add_residual`. Integrating it into the transforms would be difficult as there are many different ways the residual samples are produced before this, e.g. using core transforms, LFNST, transform skip, etc.

> > At the moment, this is manifesting in an intermittent failure for the ITX checkasm of ff_vvc_inv_dct2_dct2_32x32_10_avx2
>
> why overflow only happens at 32x32? will  16x16 .. 2x2 has the same issue? will 8 bits has the same issue?

Larger transforms produce larger residual samples simply because there are more terms in the dot products making up the matrix multiplication. The 8-bit 32x32 transform is not affected as its `bd_shift` is larger, enough to guarantee the samples remain below `INT16_MAX`. I have not yet implemented the 10-bit 64x64 transform, though I am fairly certain it will suffer from the same issue. I also think the non-extended-precision 12-bit+ 32x32 and 64x64 transforms are affected. There is also the possibility that the 8-bit 64x64 transform is affected and that smaller non-extended-precision 12-bit+ transforms are affected.

This is not really an issue in any sensible video — why would the residual ever be many times larger than the full range of a pixel? — hence not being caught by any of the conformance tests.

Reply:
> Ah I see. dav1d uses their coeff/pixel types in their signatures, and I thought this was a template which was instantiated for each bit depth, but it looks as though it is only a build-time selection.

the dav1d uses templates at a higher level. all bytefn(fn) are template functions.

>The 8-bit 32x32 transform is not affected as its bd_shift is larger, enough to guarantee the samples remain below INT16_MAX
If you check this line https://github.com/ffvvc/FFmpeg/blob/main/libavcodec/vvc/vvc_intra.c#L470
It means the input depth is 5 + sps->log2_transform_range, the output depth is sps->bit_depth. 
So the itx's output depth should be 5 + sps->log2_transform_range, not related to bit depth.
the range of sps->log_transform_range is [15, 20], so 5 + sps->log2_transform_range is [20, 25].
we can't hold it in int16_t. we need int32_t in any case.